sdk/python/kfp/local/graph_utils.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2024 The Kubeflow Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Graph algorithms which are useful for working with PipelineSpec."""
+
+from typing import Dict, List, Set
+
+from kfp.pipeline_spec import pipeline_spec_pb2
+
+
+def topological_sort_tasks(
+        tasks: Dict[str, pipeline_spec_pb2.PipelineTaskSpec]) -> List[str]:
+    """Given a dictionary of task name to PipelineTaskSpec, obtains a
+    topologically sorted stack of task names.
+
+    Args:
+        tasks: The tasks in the pipeline.
+
+    Returns:
+        A totally ordered stack of tasks. Tasks should be executed in the order they are popped off the right side of the stack.
+    """
+    adj_list = build_adjacency_list(tasks)
+    return topological_sort(adj_list)
+
+
+def build_adjacency_list(
+    tasks: Dict[str,
+                pipeline_spec_pb2.PipelineTaskSpec]) -> Dict[str, List[str]]:
+    """Builds an adjacency list from a dictionary of task name to
+    PipelineTaskSpec. This is a data structure simplification step, which
+    allows for a general topological_sort sort implementation.
+
+    Args:
+        tasks: The tasks in the pipeline.
+
+    Returns:
+        An adjacency list of tasks name to a list of upstream tasks. The key task depends on all value tasks being executed first.
+    """
+    return {
+        task_name: task_details.dependent_tasks
```

Comment:
nit: This seems to be reversed order of key and value comparing to traditional adjacency list definition. Though you get the result right by reserving the sorted result at the end.